### PR TITLE
Swift: Imprecise Taint Flows

### PR DESCRIPTION
**Pull request by geoffw0**
Adds a small number of imprecise taint flow models.  There's a *lot* more we could potentially do here, but lets start small for now:
- taint flow through methods that look like an "append" or "insert"; many of these will be on classes derived from `Collection` (which we can model and I think do model most of), but quite often these patterns are also copied in classes that do not derived from `Collection`.  By using an imprecise model, we can catch the latter as well.
- taint flow through initializers with a "data" argument, which I'm pretty happy will inherit taint from an argument of that nature.

There's a contents / flow paths issue here - we only have flow path support on MAD flow models, not on QL flow models which these imprecise sinks have to be - so we can't have flow from the contents of the argument to the contents of the qualifier.  That's why we don't catch the cases on lines 118 and 119 of the test where taint was introduced in the contents of the `taintedArray` argument rather than in `taintedArray` itself.  I'm keen to hear thoughts on how we could solve this.

### CODE REVIEW

**Comment by geoffw0** [comment]
DCA run looks clean.

File: swift/ql/lib/codeql/swift/frameworks/Heuristic.qll [hunk 1]
```diff
@@ -51,3 +52,60 @@ private class InitializerContentsOfLocalSource extends LocalFlowSource {
...
+ * ```
+ */
+private class AppendCallStep extends AdditionalTaintStep {
+  override predicate step(DataFlow::Node node1, DataFlow::Node node2) {
+    exists(CallExpr ce, Argument arg |
+      ce.getAnArgument() = arg and
+      ce.getStaticTarget().(Function).getShortName() = ["append", "insert"] and
+      arg.getLabel() = ["", "contentsOf"] and
+      node1.asExpr() = arg.getExpr() and
+      node2.asExpr() = ce.getQualifier()
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/frameworks/Heuristic.qll:89]: Should this really be the qualifier expression? I would have expected this to be `node.(PostUpdateNode).getPreUpdateNode().asExpr() = ce.getQualifier()`?

**Review by MathiasVP — COMMENTED** [review]
One comment, but otherwise this LGTM

**Comment by geoffw0** [comment]
DCA looks good to me.

**Comment by MathiasVP** [comment]
> DCA looks good to me.

I agree. I think the only remaining blocker is my comment [here](https://github.com/github/codeql/pull/14925#discussion_r1408068588) 🙂.

**Comment by geoffw0** [comment]
Sorry.  I'll get around to that change soon...

**Comment by geoffw0** [comment]
I think this is ready to merge.

**Review by MathiasVP — APPROVED** [review]
Yep. Thanks for addressing my comments!

**Pull request merged**